### PR TITLE
Metric callback improve

### PR DESCRIPTION
**Pull request by TingDaoK**
Move the telemetry callback to be invoked from the meta request event loop

- Previously, we invoke the telemetry callback once the send_data gets clean up, which happens when the s3 request gone, or the send_data get cleaned up.
- Now, we schedule the callback to be invoked from the events on the meta request
  - When the send_data get set up, which will happen if the request prepared the http message (before the signing).
  - When the finished_request invoked, which is the request will not be retried anymore and finished.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_auto_ranged_put.c [hunk 1]
```diff
@@ -1341,6 +1341,7 @@ static void s_s3_auto_ranged_put_request_finished(
             /* BEGIN CRITICAL SECTION */
             {
                 aws_s3_meta_request_lock_synced_data(meta_request);
+                aws_s3_request_finish_up_metrics_synced(request, meta_request);
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/s3_auto_ranged_put.c:1344]: It's basically every implementation of `finished_request`, which only be invoked `when the request is done being sent, and will not be retried/sent again`.

We also invoke the callback from [here](https://github.com/awslabs/aws-c-s3/pull/390/files#diff-7a5004097aea204a541fbae19de29f50a1005040ac7c70182964c5dd02ef5d78R54-R71), which is the case when the request gets retried for any reasons.

So, we cannot just queue the metric from http completion callback, as the request could end up gets retry afterward.

File: source/s3_util.c [hunk 2]
```diff
@@ -623,3 +626,30 @@ int aws_s3_crt_error_code_from_server_error_code_string(struct aws_byte_cursor e
...
+    AWS_PRECONDITION(meta_request);
+    AWS_PRECONDITION(request);
+
+    if (request->send_data.metrics != NULL) {
+        /* Request is done, complete the metrics for the request now. */
+        struct aws_s3_request_metrics *metric = request->send_data.metrics;
+        metric->crt_info_metrics.error_code = error_code;
+        if (record_end) {
+            aws_high_res_clock_get_ticks((uint64_t *)&metric->time_metrics.end_timestamp_ns);
+            metric->time_metrics.total_duration_ns =
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/s3_util.c:644]: The idea is when we have body to be delivered, we delay the end_timestamp until the callback is scheduled to be invoked, from [here](https://github.com/awslabs/aws-c-s3/pull/390/files#diff-436560b33218855638d58d1c14ecca658afd2c8ec65f2b525e21c6d829a7068aR1751-R1755)

File: source/s3_meta_request.c [hunk 3]
```diff
@@ -1711,6 +1711,22 @@ static void s_s3_meta_request_event_delivery_task(struct aws_task *task, void *a
                 aws_atomic_fetch_sub(&client->stats.num_requests_streaming_response, 1);
 
                 ++num_parts_delivered;
+
+                if (request->send_data.metrics != NULL) {
+                    /* Request is done streaming the body, complete the metrics for the request now. */
+                    struct aws_s3_request_metrics *metrics = request->send_data.metrics;
+                    metrics->crt_info_metrics.error_code = error_code;
```
Comment (graebm) [inline source/s3_meta_request.c:1718]: why are you altering the metric's error_code? the `error_code` on the right-hand-side is whether the meta-request has failed. I would think the metrics callback would want to report whether that individual HTTP request failed.

**Review by graebm — APPROVED** [review]
fix & ship

**Pull request merged**